File: learning/zig/src/vec3.zig
const std = @import("std");
const expect = std.testing.expect;
const sqrt = std.math.sqrt;

pub const Vec3 = struct {
    x: f32,
    y: f32,
    z: f32,

    fn add(self: Vec3, other: Vec3) Vec3 {
        return Vec3.new(self.x + other.x, self.y + other.y, self.z + other.z);
    }

    fn add_assign(self: *Vec3, other: Vec3) void {
        self.x += other.x;
        self.y += other.y;
        self.z += other.z;
    }

    fn div(self: Vec3, d: f32) Vec3 {
        const S = 1.0 / d;

        return Vec3.new(self.x * S, self.y * S, self.z * S);
    }

    fn div_assign(self: *Vec3, d: f32) void {
        const S = 1.0 / d;

        self.x *= S;
        self.y *= S;
        self.z *= S;
    }

    fn equals(self: Vec3, other: Vec3) bool {
        return self.x == other.x and
            self.y == other.y and
            self.z == other.z;
    }

    fn get(self: Vec3, idx: u8) Vec3Error!f32 {
        return switch (idx) {
            0 => self.x,
            1 => self.y,
            2 => self.z,
            else => Vec3Error.IndexOutOfBounds,
        };
    }

    fn mag(self: Vec3) f32 {
        return sqrt(self.x * self.x +
            self.y * self.y +
            self.z * self.z);
    }

    fn mult(self: Vec3, s: f32) Vec3 {
        return Vec3.new(
            self.x * s,
            self.y * s,
            self.z * s,
        );
    }

    fn mult_assign(self: *Vec3, s: f32) void {
        self.x *= s;
        self.y *= s;
        self.z *= s;
    }

    fn neg(self: Vec3) Vec3 {
        return Vec3.new(-self.x, -self.y, -self.z);
    }

    pub fn new(x: f32, y: f32, z: f32) Vec3 {
        return Vec3{ .x = @as(f32, x), .y = @as(f32, y), .z = @as(f32, z) };
    }

    fn norm(self: Vec3) Vec3 {
        return self.div(self.mag());
    }

    fn sub(self: Vec3, other: Vec3) Vec3 {
        return Vec3.new(self.x - other.x, self.y - other.y, self.z - other.z);
    }

    fn sub_assign(self: *Vec3, other: Vec3) void {
        self.x -= other.x;
        self.y -= other.y;
        self.z -= other.z;
    }
};

const Vec3Error = error{IndexOutOfBounds};

const EPS_F32: f32 = 1e-6;

fn approx_f32(a: f32, b: f32) bool {
    const d = a - b;
    return if (d < 0) -d < EPS_F32 else d < EPS_F32;
}

test "add" {
    const V1 = Vec3.new(1, 2, 3);
    const V2 = Vec3.new(4, 5, 6);
    const EXPECTED = Vec3.new(5, 7, 9);

    try expect(EXPECTED.equals(V1.add(V2)));
}

test "add_assign" {
    var v1 = Vec3.new(1, 2, 3);
    const V2 = Vec3.new(4, 5, 6);
    const EXPECTED = Vec3.new(5, 7, 9);

    v1.add_assign(V2);

    try expect(EXPECTED.equals(v1));
}

test "div" {
    const V = Vec3.new(2, 4, 6);
    const EXPECTED = Vec3.new(1, 2, 3);

    try expect(EXPECTED.equals(V.div(2)));
}

test "div_assign" {
    var v = Vec3.new(2, 4, 6);
    const EXPECTED = Vec3.new(1, 2, 3);

    v.div_assign(2);

    try expect(EXPECTED.equals(v));
}

test "get" {
    const V = Vec3.new(1, 2, 3);
    const EXPECTED = Vec3.new(try V.get(0), try V.get(1), try V.get(2));

    try expect(EXPECTED.equals(V));
}

test "mag" {
    const V = Vec3.new(1, 2, 2);

    try expect(approx_f32(V.mag(), 3));
}

test "mult" {
    const V = Vec3.new(1, 2, 3);
    const EXPECTED = Vec3.new(2, 4, 6);

    try expect(EXPECTED.equals(V.mult(2)));
}

test "mult_assign" {
    var v = Vec3.new(1, 2, 3);
    const EXPECTED = Vec3.new(2, 4, 6);

    v.mult_assign(2);

    try expect(EXPECTED.equals(v));
}

test "neg" {
    const V = Vec3.new(1, 2, 3);
    const EXPECTED = Vec3.new(-1, -2, -3);

    try expect(EXPECTED.equals(V.neg()));
}

test "norm" {
    const V = Vec3.new(3, 0, 0);
    const N = V.norm();

    try expect(approx_f32(N.mag(), 1));
}

test "sub" {
    const V1 = Vec3.new(5, 7, 9);
    const V2 = Vec3.new(4, 5, 6);
    const EXPECTED = Vec3.new(1, 2, 3);

    try expect(EXPECTED.equals(V1.sub(V2)));
}

test "sub_assign" {
    var v1 = Vec3.new(5, 7, 9);
    const V2 = Vec3.new(4, 5, 6);
    const EXPECTED = Vec3.new(1, 2, 3);

    v1.sub_assign(V2);

    try expect(EXPECTED.equals(v1));
}

File: learning/rust/src/vec3.rs
use std::ops::{Add, AddAssign, Neg, Sub, SubAssign};

#[derive(Debug, PartialEq)]
struct Vec3 {
    x: f32,
    y: f32,
    z: f32,
}

impl Vec3 {
    fn mag(&self) -> f32 {
        (
            self.x * self.x +
            self.y * self.y +
            self.z * self.z
        ).sqrt()
    }
    
    fn new(x: f32, y: f32, z: f32) -> Self {
        Self { x, y, z }
    }

    fn norm(&self) -> Self {
        self.scalar_div(self.mag())
    }

    fn scalar_div(&self, d: f32) -> Self {
        let s = 1.0 / d;

        Self {
            x: self.x * s,
            y: self.y * s,
            z: self.z * s,
        }   
    }

    fn scalar_div_assign(&mut self, d: f32) {
        let s = 1.0 / d;

        self.x *= s;
        self.y *= s;
        self.z *= s;
    }

    fn scalar_mult(&self, s: f32) -> Self {
        Self {
            x: self.x * s,
            y: self.y * s,
            z: self.z * s,
        }
    }

    fn scalar_mult_assign(&mut self, s: f32) {
        self.x *= s;
        self.y *= s;
        self.z *= s;
    }
}

impl Add for Vec3 {
    type Output = Self;

    fn add(self, other: Self) -> Self {
        Self {
            x: self.x + other.x,
            y: self.y + other.y,
            z: self.z + other.z,
        }
    }
}

impl AddAssign for Vec3 {
    fn add_assign(&mut self, other: Self) {
        self.x += other.x;
        self.y += other.y;
        self.z += other.z;
    }
}

impl Neg for Vec3 {
    type Output = Self;
    
    fn neg(self) -> Self {
        Self {
            x: -self.x,
            y: -self.y,
            z: -self.z,
        }
    }
}

impl Sub for Vec3 {
    type Output = Self;

    fn sub(self, other: Self) -> Self {
        Self {
            x: self.x - other.x,
            y: self.y - other.y,
            z: self.z - other.z,
        }
    }
}

impl SubAssign for Vec3 {
    fn sub_assign(&mut self, other: Self) {
        self.x -= other.x;
        self.y -= other.y;
        self.z -= other.z;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const EPS_F32: f32 = 1e-6;

    fn approx_f32(a: f32, b: f32) -> bool {
        let d = a - b;

        match d < 0.0 {
            true => { -d < EPS_F32 },
            false => { d < EPS_F32 }
        }
    }

    #[test]
    fn test_add() {
        let v1 = Vec3::new(1.0, 2.0, 3.0);
        let v2 = Vec3::new(4.0, 5.0, 6.0);
        let expected = Vec3::new(5.0, 7.0, 9.0);

        assert_eq!(expected, v1 + v2);
    }

    #[test]
    fn test_add_assign() {
        let mut v1 = Vec3::new(1.0, 2.0, 3.0);
        let v2 = Vec3::new(4.0, 5.0, 6.0);
        let expected = Vec3::new(5.0, 7.0, 9.0);

        v1 += v2;

        assert_eq!(expected, v1);
    }

    #[test]
    fn test_mag() {
        let v = Vec3::new(1.0, 2.0, 2.0);

        assert!(approx_f32(v.mag(), 3.0));
    }

    #[test]
    fn test_neg() {
        let v = Vec3::new(1.0, 2.0, 3.0);
        let expected = Vec3::new(-1.0, -2.0, -3.0);

        assert_eq!(expected, -v);
    }

    #[test]
    fn test_norm() {
        let v = Vec3::new(3.0, 0.0, 0.0);
        let n = v.norm();

        assert!(approx_f32(n.mag(), 1.0));
    }


    #[test]
    fn test_scalar_div() {
        let v = Vec3::new(2.0, 4.0, 6.0);
        let expected = Vec3::new(1.0, 2.0, 3.0);

        assert_eq!(expected, v.scalar_div(2.0));
    }

    #[test]
    fn test_scalar_div_assign() {
        let mut v = Vec3::new(2.0, 4.0, 6.0);
        let expected = Vec3::new(1.0, 2.0, 3.0);

        v.scalar_div_assign(2.0);

        assert_eq!(expected, v);
    }
    
    #[test]
    fn test_scalar_mult() {
        let v = Vec3::new(1.0, 2.0, 3.0);
        let expected = Vec3::new(2.0, 4.0, 6.0);

        assert_eq!(expected, v.scalar_mult(2.0));
    }

    #[test]
    fn test_scalar_mult_assign() {
        let mut v = Vec3::new(1.0, 2.0, 3.0);
        let expected = Vec3::new(2.0, 4.0, 6.0);

        v.scalar_mult_assign(2.0);

        assert_eq!(expected, v);
    }

    #[test]
    fn test_sub() {
        let v1 = Vec3::new(5.0, 7.0, 9.0);
        let v2 = Vec3::new(4.0, 5.0, 6.0);
        let expected = Vec3::new(1.0, 2.0, 3.0);

        assert_eq!(expected, v1 - v2);
    }

    #[test]
    fn test_sub_assign() {
        let mut v1 = Vec3::new(5.0, 7.0, 9.0);
        let v2 = Vec3::new(4.0, 5.0, 6.0);
        let expected = Vec3::new(1.0, 2.0, 3.0);

        v1 -= v2;

        assert_eq!(expected, v1);
    }
}
